### ISSUE TITLE
Double reporting tests on dart2js-win7-ie10chrome

### ISSUE THREAD

**Issue by peter-ahe-google** [dart-lang/sdk#28873]
The following co19 tests are double reporting on dart2js-win7-ie10chrome:

* Language/Expressions/Function_Invocation/async_cleanup_t01
* Language/Expressions/Function_Invocation/async_cleanup_t03
* Language/Expressions/Function_Invocation/async_cleanup_t05
* Language/Expressions/Function_Invocation/async_cleanup_t06
* Language/Expressions/Function_Invocation/async_cleanup_t08
* Language/Statements/Yield_and_Yield_Each/Yield/execution_async_t03
* Language/Statements/Yield_and_Yield_Each/Yield/execution_async_t05
* Language/Statements/Yield_and_Yield_Each/Yield/execution_async_t06
* Language/Statements/Yield_and_Yield_Each/Yield/execution_async_t08
* LayoutTests/fast/dom/HTMLLinkElement/link-onload-stylesheet-with-import_t01
* LayoutTests/fast/dom/mutation-event-remove-inserted-node_t01
* LayoutTests/fast/dom/shadow/event-path-not-in-document_t01
* LayoutTests/fast/forms/search-popup-crasher_t01
* LayoutTests/fast/loader/stateobjects/replacestate-in-onunload_t01
* LayoutTests/fast/replaced/preferred-widths_t01
* LayoutTests/fast/replaced/table-percent-width_t01
* LayoutTests/fast/table/hittest-tablecell-with-borders-right-edge_t01
* LayoutTests/fast/table/table-colgroup-present-after-table-row_t01
* LayoutTests/fast/table/table-with-content-width-exceeding-max-width_t01
* LayoutTests/fast/text/find-case-folding_t01
* LayoutTests/fast/text/zero-width-characters-complex-script_t01.dart
* LayoutTests/fast/transforms/hit-test-large-scale_t01
* LibTest/html/Window/postMessage_A01_t01
* WebPlatformTest/dom/EventTarget/dispatchEvent_A02_t01

It's unclear if these tests are broken, of if this is a problem in dart2js. I'm assigning this to dart2js for initial evaluation.

In case this is a bug in either dart2js or the tests, I've skipped them on all dart2js configurations. Please adjust the status file after investigating.

**Comment by peter-ahe-google** [comment]
@sgrekhov could this be a problem in the tests? Double reporting means that a test reports success or failure before all asynchronous operations have completed. In most cases, tests should be written in such a way that this can't happen, even if something fails. In practice, a double reporting test means that someone forgot to "await" a future or install an error handler on a stream.

**Comment by peter-ahe-google** [comment]
Updated list above with more tests found to be double reporting in [dart2js-mac10.11-safari-1-3-be build #3915](https://uberchromegw.corp.google.com/i/client.dart/builders/dart2js-mac10.11-safari-1-3-be/builds/3915/steps/dart2js%20safari%20co19%20tests/logs/stdio/text).

**Comment by peter-ahe-google** [comment]
Additional tests found double reporting in (dart2js-linux-chromeff-4-4-be build #3438)[https://uberchromegw.corp.google.com/i/client.dart/builders/dart2js-linux-chromeff-4-4-be/builds/3438/steps/dart2js%20ff%20co19%20tests/logs/stdio].

* LayoutTests/fast/dom/subtree-modified-attributes_t01
* LayoutTests/fast/table/css-table-max-height_t01
* LayoutTests/fast/table/table-size-integer-overflow_t01
* LayoutTests/fast/transforms/scrollIntoView-transformed_t01

**Comment by sgrekhov** [comment]
@peter-ahe-google we are working on that

**Comment by peter-ahe-google** [comment]
@sgrekhov thank you!

**Comment by eernstg** [comment]
More double reporting tests:
```
LayoutTests/fast/text/international/cjk-segmentation_t01: Skip
LayoutTests/fast/text/find-quotes_t01: Skip
LayoutTests/fast/text/find-spaces_t01: Skip
LayoutTests/fast/text/find-hidden-text_t01: Skip
LayoutTests/fast/files/file-reader-done-reading-abort_t01: Skip
LayoutTests/fast/xpath/ambiguous-operators_t01: Skip
LayoutTests/fast/xpath/4XPath/Core/test_core_functions_t02: Skip
LayoutTests/fast/mediastream/getusermedia_t01: Skip
LayoutTests/fast/shapes/shape-outside-floats/shape-outside-floats-image-margin_t01: Skip
LayoutTests/fast/writing-mode/percentage-margins-absolute_t01: Skip
LayoutTests/fast/media/mq-parsing_t01: Skip
LayoutTests/fast/xmlhttprequest/xmlhttprequest-responsetype-abort_t01: Skip
LayoutTests/fast/replaced/table-percent-height_t01: Skip
LayoutTests/fast/table/table-all-rowspans-height-distribution-in-rows-except-overlapped_t01: Skip
LayoutTests/fast/table/hittest-tablecell-right-edge_t01: Skip
LayoutTests/fast/table/table-rowspan-height-distribution-in-rows_t02: Skip
```

**Comment by alsemenov** [comment]
I fixed suspicious code in Language tests. Regarding remaining tests I have no idea how test result can be reported more than once. Could you provide more details about how it happens? Also could you try to run fixed tests? Whether the problem is really fixed there.

**Comment by eernstg** [comment]
Yet another double reporting test, reported in [this log file from dart2js-win7-ie11ff-2-4-be](https://uberchromegw.corp.google.com/i/client.dart/builders/dart2js-win7-ie11ff-2-4-be/builds/3909/steps/dart2js%20ff%20co19%20fast-startup%20tests/logs/stdio):
```
LayoutTests/fast/css/font-face-cache-bug_t01.dart: Skip
```
I don't know how this result can be reported more than once (my first assumption is "some expression `e` should have been `await e`, possibly as an expressionStatement where it doesn't matter that it evaluates to a `Future<T>` rather than a `T` for some `T`, but I suppose that's your first thought as well). I suspect the 'double reporting' flag may also be attached to the wrong test, because it is a result from an earlier test that pops up.

In any case, the first step would be to note the problem and update the status file such that it doesn't shadow other issues. So I'm updating co19-dart2js.status as shown above, and then I hope somebody else can be more helpful.

**Comment by peter-ahe-google** [comment]
It is possible that this is caused by a new test that's interfering with following tests.

In case you didn't know, when we link to the build bot, we often use a private internal URL (which is a mistake, that unfortunately is hard to avoid). Fortunately, the private URL can be translated to a public URL, by replacing the first part up to (and including `/p/client.dart/`) with `https://build.chromium.org/p/client.dart/`.

Here are the ones Erik added:

https://build.chromium.org/p/client.dart/builders/dart2js-win7-ie11ff-2-4-be/builds/3911/steps/dart2js%20ff%20co19%20tests/logs/stdio

https://build.chromium.org/p/client.dart/builders/dart2js-win7-ie11ff-2-4-be/builds/3909/steps/dart2js%20ff%20co19%20fast-startup%20tests/logs/stdio

I'm updating my own messages above to use the public URL.

**Comment by alsemenov** [comment]
Could you provide the exact output of some single double-reporting test?

**Comment by peter-ahe-google** [comment]
I'm not sure we have that information in our logs, @whesse what are your thought about getting more information about these double reporting tests?

**Comment by alsemenov** [comment]
So far I am trying to understand how the double-reporting could happen.  Most of the tests are asynchronous. All tests use common library for signalling asynchronous result according to http://code.google.com/p/co19/issues/detail?id=423 
Therefore I can imagine the following possible scenarios:
1. `asyncStart()` is called from async function, which is executed after `main()` finishes. So test runner recognizes the test as synchronous, but it also reports asynchronous results
2. There are some unexpected events which cause `asyncEnd()` to be called more times than `asyncStart()` was called. In this case the library produces an error, which may be considered as double reporting by test runner. 

I already have fixed first case for Language tests. However this is not the case for the rest tests. So probably the problem is caused by second case or due to some other reason (for example some bug in test runner). Exact output of the tests would be very helpful for analysis.

**Comment by peter-ahe-google** [comment]
I agree. We need to provide you with more information.

**Comment by peter-ahe-google** [comment]
I've assigned @whesse to make it clear that we're waiting for a reply to the above question about his thoughts on getting more information.

**Comment by whesse** [comment]
Do we know if all or most of the double-reporting tests are passing or failing?
Could there be any printing of "pass" or "fail", or is it very likely the calls to the asyncStart and asyncEnd functions that are the root cause?

I will look into this.

**Comment by peter-ahe-google** [comment]
Thank you, @whesse. Let's talk about this face to face when you're back.

**Comment by alsemenov** [comment]
Along with @sgrekhov we managed to run and analyse the double reporting tests.

There is at least two group of tests, that are double reporting due to two different reasons:

1. `asyncStart()` is called from async function, which is executed after main() finishes. Already fixed in 	33ec795 and in 490beda

2. compiled javascript test code contains errors. For example (I formatted the output for ease of reading):

> 2017-03-10 16:35:12.512750 Warning: Double reporting tests:
> 2017-03-10 16:35:12.512750 Warning: http://127.0.0.1:54842/root_build/generated_tests/dart2js-ff-sdk/short6412_age-2d-with-video-rgba4444_t01/test.html?crossOriginPort=54843, output: 
> 2017-03-10 16:35:12.512750 Warning: 
> [
> 	{"type":"debug","value":"test_controller.js started","timestamp":0},
> 	{"type":"debug","value":"Sending events to driver page (isFirstMessage = true, isStatusUpdate = true, isDone = false)","timestamp":0},
> 	{"type":"print","value":"dart-calling-main","timestamp":0.319},
> 	{"type":"print","value":"unittest-suite-wait-for-done","timestamp":0.354},
> 	{"type":"print","value":"dart-main-done","timestamp":0.355},
> 	{"type":"message_received","value":"dart-calling-main","timestamp":0.946},
> 	{"type":"message_received","value":"unittest-suite-wait-for-done","timestamp":0.947},
> 	{"type":"message_received","value":"dart-main-done","timestamp":0.947},
> 	{"type":"print","value":"unittest-suite-success","timestamp":2.647},
> 	{"type":"message_received","value":"unittest-suite-success","timestamp":2.725},
> 	{"type":"debug","value":"Sending events to driver page (isFirstMessage = false, isStatusUpdate = false, isDone = true)","timestamp":2.726},
> 	**{"type":"window_onerror","value":"window.onerror called: \n\nhttp://127.0.0.1:54842/root_build/generated_compilations/dart2js-sdk/short6413_age-2d-with-video-rgba4444_t01/test.js:2262:\nError: \n\n","timestamp":8.342},**
> 	{"type":"debug","value":"Sending events to driver page (isFirstMessage = false, isStatusUpdate = false, isDone = true)","timestamp":8.343},
> 	{"type":"dom","value":"<head>\n  <meta http-equiv=\"X-UA-Compatible\" content=\"IE=edge\">\n  <meta name=\"dart.unittest\" content=\"full-stack-traces\">\n  <title> Test /D:/Google/dart/dart-sdk/sdk/tests/co19/src/LayoutTests/fast/canvas/webgl/tex-image-and-sub-image-2d-with-video-rgba4444_t01.dart </title>\n  <style>\n     .unittest-table { font-family:monospace; border:1px; }\n     .unittest-pass { background: #6b3;}\n     .unittest-fail { background: #d55;}\n     .unittest-error { background: #a11;}\n  </style>\n</head>\n<body>      <canvas id=\"example\" width=\"32px\" height=\"32px\"></canvas>\n      <div id=\"console\"><span>Verify texImage2D and texSubImage2D code paths taking video elements (6408/32819)<br></span><span>Testing texImage2D with flipY=true<br></span><span>Checking lower left corner<br></span><span>shouldBe [0, 255, 0]<br></span><span>Checking upper left corner<br></span><span>shouldBe [255, 0, 0]<br></span><span>Testing texImage2D with flipY=false<br></span><span>Checking lower left corner<br></span><span>shouldBe [255, 0, 0]<br></span><span>Checking upper left corner<br></span><span>shouldBe [0, 255, 0]<br></span><span>Testing texSubImage2D with flipY=true<br></span><span>Checking lower left corner<br></span><span>shouldBe [0, 255, 0]<br></span><span>Checking upper left corner<br></span><span>shouldBe [255, 0, 0]<br></span><span>Testing texSubImage2D with flipY=false<br></span><span>Checking lower left corner<br></span><span>shouldBe [255, 0, 0]<br></span><span>Checking upper left corner<br></span><span>shouldBe [0, 255, 0]<br></span><span>getError was expected value: NO_ERROR : should be no errors<br></span><span>Testing texImage2D with flipY=true<br></span><span>Checking lower left corner<br></span><span>shouldBe [0, 255, 0]<br></span><span>Checking upper left corner<br></span><span>shouldBe [255, 0, 0]<br></span><span>Testing texImage2D with flipY=false<br></span><span>Checking lower left corner<br></span><span>shouldBe [255, 0, 0]<br></span><span>Checking upper left corner<br></span><span>shouldBe [0, 255, 0]<br></span><span>Testing texSubImage2D with flipY=true<br></span><span>Checking lower left corner<br></span><span>shouldBe [0, 255, 0]<br></span><span>Checking upper left corner<br></span><span>shouldBe [255, 0, 0]<br></span><span>Testing texSubImage2D with flipY=false<br></span><span>Checking lower left corner<br></span><span>shouldBe [255, 0, 0]<br></span><span>Checking upper left corner<br></span><span>shouldBe [0, 255, 0]<br></span><span>getError was expected value: NO_ERROR : should be no errors<br></span></div>\n      <video id=\"vid\" controls=\"\" loop=\"\" width=\"640\" height=\"228\">\n        <source src=\"/root_dart/tests/co19/src/LayoutTests/fast/canvas/webgl/resources/red-green.mp4\" type=\"video/mp4; codecs=&quot;avc1.42E01E, mp4a.40.2&quot;\">\n        <source src=\"/root_dart/tests/co19/src/LayoutTests/fast/canvas/webgl/resources/red-green.webmvp8.webm\" type=\"video/webm; codecs=&quot;vp8, vorbis&quot;\">\n        <source src=\"/root_dart/tests/co19/src/LayoutTests/fast/canvas/webgl/resources/red-green.theora.ogv\" type=\"video/ogg; codecs=&quot;theora, vorbis&quot;\">\n      </video>\n      <canvas width=\"16\" height=\"16\" style=\"border: 1px solid black;\"></canvas><canvas width=\"16\" height=\"16\" style=\"border: 1px solid black;\"></canvas><canvas width=\"16\" height=\"16\" style=\"border: 1px solid black;\"></canvas><canvas width=\"16\" height=\"16\" style=\"border: 1px solid black;\"></canvas><canvas width=\"16\" height=\"16\" style=\"border: 1px solid black;\"></canvas><canvas width=\"16\" height=\"16\" style=\"border: 1px solid black;\"></canvas><canvas width=\"16\" height=\"16\" style=\"border: 1px solid black;\"></canvas><canvas width=\"16\" height=\"16\" style=\"border: 1px solid black;\"></canvas></body>","timestamp":8.343}
> ]

Another example:

> 2017-03-10 16:35:12.512750 Warning: 
> 2017-03-10 16:35:12.512750 Warning: http://127.0.0.1:54842/root_build/generated_tests/dart2js-ff-sdk/co19_caretrangefrompoint_in-zoom-and-scroll_t01/test.html?crossOriginPort=54843, output: 
> 2017-03-10 16:35:12.512750 Warning: 
> [
> 	{"type":"debug","value":"test_controller.js started","timestamp":0},
> 	{"type":"debug","value":"Sending events to driver page (isFirstMessage = true, isStatusUpdate = true, isDone = false)","timestamp":0.001},
> 	{"type":"print","value":"dart-calling-main","timestamp":0.708},
> 	{"type":"print","value":"unittest-suite-wait-for-done","timestamp":1.03},
> 	**{"type":"window_onerror","value":"window.onerror called: \n\nhttp://127.0.0.1:54842/root_build/generated_compilations/dart2js-sdk/co19_caretrangefrompoint_in-zoom-and-scroll_t01/test.js:6293:\nTypeError: $event.get$clientX is not a function\n\n","timestamp":1.215},**
> 	{"type":"debug","value":"Sending events to driver page (isFirstMessage = false, isStatusUpdate = false, isDone = true)","timestamp":1.216},
> 	**{"type":"window_onerror","value":"window.onerror called: \n\nhttp://127.0.0.1:54842/root_build/generated_compilations/dart2js-sdk/co19_caretrangefrompoint_in-zoom-and-scroll_t01/test.js:6293:\nTypeError: $event.get$clientX is not a function\n\n","timestamp":1.254},**
> 	{"type":"debug","value":"Sending events to driver page (isFirstMessage = false, isStatusUpdate = false, isDone = true)","timestamp":1.255},
> 	**{"type":"window_onerror","value":"window.onerror called: \n\nhttp://127.0.0.1:54842/root_build/generated_compilations/dart2js-sdk/co19_caretrangefrompoint_in-zoom-and-scroll_t01/test.js:6293:\nTypeError: $event.get$clientX is not a function\n\n","timestamp":1.258},**
> 	{"type":"debug","value":"Sending events to driver page (isFirstMessage = false, isStatusUpdate = false, isDone = true)","timestamp":1.259},
> 	**{"type":"window_onerror","value":"window.onerror called: \n\nhttp://127.0.0.1:54842/root_build/generated_compilations/dart2js-sdk/co19_caretrangefrompoint_in-zoom-and-scroll_t01/test.js:6293:\nTypeError: $event.get$clientX is not a function\n\n","timestamp":1.271},**
> 	{"type":"debug","value":"Sending events to driver page (isFirstMessage = false, isStatusUpdate = false, isDone = true)","timestamp":1.298},
> 	{"type":"dom","value":"<head>\n  <meta http-equiv=\"X-UA-Compatible\" content=\"IE=edge\">\n  <meta name=\"dart.unittest\" content=\"full-stack-traces\">\n  <title> Test /D:/Google/dart/dart-sdk/sdk/tests/co19/src/LayoutTests/fast/dom/Document/CaretRangeFromPoint/caretRangeFromPoint-in-zoom-and-scroll_t01.dart </title>\n  <style>\n     .unittest-table { font-family:monospace; border:1px; }\n     .unittest-pass { background: #6b3;}\n     .unittest-fail { background: #d55;}\n     .unittest-error { background: #a11;}\n  </style>\n</head>\n<body>    <div id=\"first\" style=\"width: 4000px; height: 1000px\">A long text to test document caretRangeFromPoint. Test scroll, zoom, zoomAndScroll .................................... The End </div>\n    </body>","timestamp":1.298}
> ]
>
